tests/test_coordinate_source.py

```diff
@@ -0,0 +1,328 @@
+            self.assertTrue(np.array_equal(imgs_box[i], imgs_coord[i]))
+            self.assertTrue(np.array_equal(imgs_coord[i], imgs_star[i]))
+            self.assertTrue(np.array_equal(imgs_star[i], imgs_relion[i]))
+
+    def testImagesRandomIndices(self):
+        # ensure that we can load a specific, possibly out of order, list of
+        # indices, and that the result is in the order we asked for
+        images_in_order = self.src_from_box.images(0, 440)
+        # test loading every other image and compare
+        odd = np.array([i for i in range(1, 440, 2)])
+        even = np.array([i for i in range(0, 439, 2)])
+        odd_images = self.src_from_box._images(indices=odd)
+        even_images = self.src_from_box._images(indices=even)
+        for i in range(0, 220):
+            self.assertTrue(np.array_equal(images_in_order[2 * i], even_images[i]))
+            self.assertTrue(np.array_equal(images_in_order[2 * i + 1], odd_images[i]))
+
+        # random sample of [0,440) of length 100
+        random_sample = np.array(random.sample([i for i in range(440)], 100))
+        random_images = self.src_from_box._images(indices=random_sample)
+        for i, idx in enumerate(random_sample):
+            self.assertTrue(np.array_equal(images_in_order[idx], random_images[i]))
+
+    def testMaxRows(self):
+        imgs = self.src_from_box.images(0, 440)
+        # make sure max_rows loads the correct particles
+        src_100 = EmanCoordinateSource(self.files_box, max_rows=100)
+        imgs_100 = src_100.images(0, src_100.n)
+        for i in range(100):
+            self.assertTrue(np.array_equal(imgs[i], imgs_100[i]))
+        # make sure max_rows > self.n loads max_rows images
+        src_500 = EmanCoordinateSource(self.files_box, max_rows=500)
+        self.assertEqual(src_500.n, 440)
+        imgs_500 = src_500.images(0, 440)
+        for i in range(440):
+            self.assertTrue(np.array_equal(imgs[i], imgs_500[i]))
+        # make sure max_rows loads correct particles
+        # when some have been excluded
+        imgs_newsize = EmanCoordinateSource(self.files_box, particle_size=336).images(
+            0, 50
+        )
+        src_maxrows = EmanCoordinateSource(
+            self.files_box, particle_size=336, max_rows=50
+        )
+        # max_rows still loads 50 images even if some particles were excluded
+        self.assertEqual(src_maxrows.n, 50)
+        imgs_maxrows = src_maxrows.images(0, 50)
+        for i in range(50):
+            self.assertTrue(np.array_equal(imgs_newsize[i], imgs_maxrows[i]))
+
+    def testBoundaryParticlesRemoved(self):
+        src_centers_larger_particles = CentersCoordinateSource(
+            self.files_coord, particle_size=300
+        )
+        src_box_larger_particles = EmanCoordinateSource(
+            self.files_box, particle_size=300
+        )
+        # 2 particles do not fit at this particle size
+        self.assertEqual(src_centers_larger_particles.n, 438)
```

Comment:
Brittle…

Reply:
We now have randomly generated particle centers, so we know from the outset that exactly 100 particles will not fit.

Reply:
:+1: